internal/fxreflect/fxreflect.go

```diff
@@ -25,25 +25,84 @@ import (
 	"reflect"
 	"runtime"
 	"strings"
+
+	"go.uber.org/dig"
 )
 
-// ReturnTypes takes a func and returns a slice of string'd types
-// TODO instead of duplicating the dig's reflect logic, trying to
-// determine which types actually made it into the container, have
-// dig return a Result struct which could contain the ProvidedTypes
+// ReturnTypes takes a func and returns a slice of string'd types.
 func ReturnTypes(t interface{}) []string {
+	// TODO(grayson): instead of duplicating the dig's reflect logic, trying to
+	// determine which types actually made it into the container, have
+	// dig return a Result struct which could contain the ProvidedTypes
+
+	if reflect.TypeOf(t).Kind() != reflect.Func {
+		// Invalid provide, will be logged as an error.
+		return []string{}
+	}
+
 	rtypes := []string{}
-	fn := reflect.ValueOf(t).Type()
+	ft := reflect.ValueOf(t).Type()
 
-	for i := 0; i < fn.NumOut(); i++ {
-		if !isErr(fn.Out(i)) {
-			rtypes = append(rtypes, fn.Out(i).String())
-		}
+	for i := 0; i < ft.NumOut(); i++ {
+		t := ft.Out(i)
+
+		traverseOuts(key{t: t}, func(s string) {
+			rtypes = append(rtypes, s)
+		})
 	}
 
 	return rtypes
 }
 
+// this type is basically straight out of dig, which is a strong signal
+// that exporting it could really DRY up some things for fx-dig relationship.
```

Comment:
Yes, that's what I'm working on prototyping now. There are several ways we can go about this, but at the very least I think exposing the `type key struct` would be a good start, as that would remove a lot of the duplicated logic here.

Communicating the keys back is going to be the tricky bit. Modifying a `Provide` API to return `([]Key, error)` is not really an option. That would break far to many places and not for that good of a reason, I think.

I think I may try to add a `Result() []Key` *method* on the container as a way to grab at the last `Provide` call.

If that's the case then we can end up with something like:

```go
c := dig.New()
err := c.Provide(func() (A,B) { ... })
// check err
for k := c.Result {
  fmt.Println("Inserted", k, "into the graph")
}
```